### PR TITLE
Prepare Data

### PR DESCRIPTION
**Pull request by reyoung**
See [here](http://ali.reyoung.me:41080/user_guides/howto/prepare_data/index.html)

### CODE REVIEW

File: source/user_guides/howto/prepare_data/index.rst [hunk 1]
```diff
@@ -0,0 +1,49 @@
+..  _user_guide_prepare_data:
+
+########
+准备数据
+########
+
+PaddlePaddle Fluid支持两种传入数据的方式： 一种用户需要使用 :code:`fluid.layers.data`
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/index.rst]: 持两种传入数据的方式， 每种方式，重新起一行好一些？

File: source/user_guides/howto/prepare_data/index.rst [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+配置数据输入层，并在 :ref:`api_guide_executor` 或 :ref:`api_guide_parallel_executor`
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/index.rst]: ref:`api_guide_parallel_executor 进来没看到东西啊？~

File: source/api_guides/low_level/executor/executor.rst [hunk 3]
```diff
@@ -15,4 +17,4 @@ Python脚本驱动 :code:`Executor` 执行。默认情况下 :code:`Executor` 
 :ref:`api_guide_low_level_program` 。
```
Comment (guochaorong) [inline source/api_guides/low_level/executor/executor.rst:17]: 这些模块 是别的老师完善么？

File: source/user_guides/howto/prepare_data/index.rst [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
++============+==================================+=======================================+
+| API接口    | :code:`executor.run(feed=...)`   |         :ref:`api_guide_reader`       |
++------------+----------------------------------+---------------------------------------+
+| 数据格式   |           Numpy Array            | :ref:`api_guide_recordio_file_format` |
++------------+----------------------------------+---------------------------------------+
+| 数据增强   | Python端使用其他库完成           | 使用Fluid中的Operator 完成            |
++------------+----------------------------------+---------------------------------------+
+|   速度     |                 慢               |                 快                    |
++------------+----------------------------------+---------------------------------------+
+| 推荐用途   |   调试模型                       |   工业训练                            |
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/index.rst:32]: read io方式  支持预测了吗~

File: source/user_guides/howto/prepare_data/feeding_data.rst [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+   label = fluid.layers.data(name="label", shape=[1], dtype="int64")
+
+   # use image/label as layer input
+   prediction = fluid.layers.fc(input=image, size=1000, act="softmax")
+   loss = fluid.layers.cross_entropy(input=prediction, label=label)
+   ...
+
+上段代码中，:code:`image` 和 :code:`label` 是通过 :code:`fluid.layers.data`
+创建的两个输入数据层。其中 :code:`image` 是 :code:`[3, 224, 224]` 维度的浮点数据;
+:code:`data` 是 :code:`[1]` 维度的整数数据。这里需要注意的是:
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/feeding_data.rst]: 这里data 是label吧

File: source/user_guides/howto/prepare_data/feeding_data.rst [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+1. Fluid中默认使用 :code:`-1` 表示 batch size 维度，默认情况下会在 :code:`shape`
+的第一个维度添加 :code:`-1` 。 所以 上段代码中， 我们可以接受将一个
+:code:`[32, 3, 224, 224]`的numpy array传给 :code:`image`。 如果想自定义batch size
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/feeding_data.rst]: 这里输出格式看着不太对， 是不是 逗号需要转移？

File: source/user_guides/howto/prepare_data/feeding_data.rst [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+维度的位置的话，请设置 :code:`fluid.layers.data(append_batch_size=False)` 。
+
+2. Fluid中目前使用 :code:`int64` 表示类别标签。
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/feeding_data.rst]: 为什么说 int64是 类别标签？ label 是类别标签？

File: source/user_guides/howto/prepare_data/feeding_data.rst [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+   ...
+
+上段代码中，:code:`image` 和 :code:`label` 是通过 :code:`fluid.layers.data`
+创建的两个输入数据层。其中 :code:`image` 是 :code:`[3, 224, 224]` 维度的浮点数据;
+:code:`data` 是 :code:`[1]` 维度的整数数据。这里需要注意的是:
+
+1. Fluid中默认使用 :code:`-1` 表示 batch size 维度，默认情况下会在 :code:`shape`
+的第一个维度添加 :code:`-1` 。 所以 上段代码中， 我们可以接受将一个
+:code:`[32, 3, 224, 224]`的numpy array传给 :code:`image`。 如果想自定义batch size
+维度的位置的话，请设置 :code:`fluid.layers.data(append_batch_size=False)` 。
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/feeding_data.rst]: 这里能否举个栗子~

File: source/user_guides/howto/prepare_data/feeding_data.rst [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+2. Fluid中目前使用 :code:`int64` 表示类别标签。
+
+传递训练数据给执行器
+####################
+
+:code:`Executor.run` 和 :code:`ParallelExecutor.run` 都接受一个 :code:`feed` 参数。
+这个参数是一个Python的字典。他的键是数据层的名字，例如上文代码中的:code:`image`。
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/feeding_data.rst]: 输出格式不对，:code:`image`  前面需要空格？

Comment (guochaorong) [inline source/user_guides/howto/prepare_data/feeding_data.rst]: ‘他的键’， 都统一用 '它' 更合适么？

File: source/user_guides/howto/prepare_data/feeding_data.rst [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+   })
+
+进阶使用
+########
+
+如何传入序列数据
+----------------
+
+序列数据是PaddlePaddle Fluid支持的特殊数据类型，可以使用 :code:`LoDTensor` 作为
+输入数据类型。它需要用户传入一个mini-batch需要被训练的所有数据和每个序列的长度信息。
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/feeding_data.rst]: 它需要用户传入一个mini-batch需要被训练的所有数据和每个序列的长度信息  需要 断下句~

File: source/user_guides/howto/prepare_data/feeding_data.rst [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+具体可以使用 :code:`fluid.create_lod_tensor` 来创建 :code:`LoDTensor`。
+
+传入序列信息的时候，需要设置序列嵌套深度，:code:`lod_level`。
+例如训练数据是词汇组成的句子，:code:`lod_level=1`；训练数据是 词汇先组成了句子，
+句子再组成了段落，那么 :code:`lod_level=2`。
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/feeding_data.rst]: 能不能加一下 lod_level=2 的栗子呀~~， 或者ref 到一个 这种栗子

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+   feeder = fluid.DataFeeder(feed_list=[img, label], place=fluid.CPUPlace())
+
+   BATCH_SIZE = 32
+   reader = paddle.batch(reader_creator(), batch_size=BATCH_SIZE)
+   fluid.recordio_writer.convert_reader_to_recordio_file(
+      "train.recordio", feeder=feeder, reader_creator=train_reader)
+
+其中 :code:`reader_creator` 创建了一个 :code:`Reader`。 :code:`fluid.DataFeeder`
+是将 :code:`Reader` 转换成 :code:`LoDTensor` 的工具。详细请参考
+:ref:`user_guide_reader` 。
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/use_recordio_reader.rst]: 感觉这里 链接fluid.DataFeeder 的api更合适？

**Comment by shanyi15** [comment]
1. 在feeding_data.rst中，
1和2数字似乎没对齐。
image

另外上图中，引代码的地方可能是没空格，和文字黏在一起了。

2. 另外，左侧导航栏的结构有点迷惑
看起来是：

准备数据
----使用Numpy Array作为训练数据
----使用RecordIO文件作为训练数据

 Python Reader

但点击“准备数据”和点击“Python Reader”，其实跳到了一个页面上（只是锚点位置不一样）

http://ali.reyoung.me:41080/user_guides/howto/prepare_data/index.html#

感觉“准备数据”和“python Reader”应该拆成两个页面，这两个页面应该分别放它们子目录的概述。

3. Python Reader的中文版会继续加在这个pr里吗？

其他没什么问题，赞

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+       return __impl__
+
+   img = fluid.layers.data(name="image", shape=[3, 224, 224])
+   label = fluid.layers.data(name="label", shape=[1], dtype="int64")
+   feeder = fluid.DataFeeder(feed_list=[img, label], place=fluid.CPUPlace())
+
+   BATCH_SIZE = 32
+   reader = paddle.batch(reader_creator(), batch_size=BATCH_SIZE)
+   fluid.recordio_writer.convert_reader_to_recordio_file(
+      "train.recordio", feeder=feeder, reader_creator=train_reader)
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/use_recordio_reader.rst]: reader_creator=train_reader， train_reader 是什么？

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+   img, label = fluid.layers.read_file(file_obj)
+
+需要注意的是，如果数据集中的最后几个样本不能组成 :code:`batch_size` 大小的批量数据，
+那么这几个样本直接组成一个批量数据进行训练。
+
+读入数据的shuffle
+-----------------
+
+使用 :ref:`api_fluid_layers_shuffle` 可以在训练过程中动态重拍训练数据。例如
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/use_recordio_reader.rst]: 重拍训练数据   重排

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+   import paddle.fluid as fluid
+   file_obj = fluid.layers.open_files(...)
+   file_obj = fliud.layers.shuffle(file_obj, buffer_size=8192)
+
+   img, label = fliud.layers.read_file(file_obj)
+
+需要注意的是:
+
+1. :code:`shuffle` 的实现先读入 :code:`buffer_size` 条样本，再随机的选出样本进行
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/use_recordio_reader.rst]: '的实现'  改成实现？

Comment (guochaorong) [inline source/user_guides/howto/prepare_data/use_recordio_reader.rst]: 每次拿出来 buffer size大小的数据到内存 重排的意思么？  “再随机的选出样本进行训练 ”  改为： 随机重排buffer size 大小的数据，进行训练，？

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+
+
+
+进阶使用
+########
+
+
+使用 :ref:`api_fluid_layers_double_buffer`
+------------------------------------------
+
```
Comment (guochaorong) [inline source/user_guides/howto/prepare_data/use_recordio_reader.rst:94]: 可以说明下 这个的使用场景 或使用效果啥的嘛

**Comment by guochaorong** [comment]
哈哈 写得非常明白， 鉴定完毕~

File: source/api_guides/low_level/executor/executor.rst (hunk 3 above)
Comment (reyoung) [reply source/api_guides/low_level/executor/executor.rst:17]: API Guide 不在 目前7月5日目标之内。。我有时间的时候，可以写一下。。

File: source/user_guides/howto/prepare_data/index.rst (hunk 4 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/index.rst:32]: 一直支持的。

File: source/user_guides/howto/prepare_data/feeding_data.rst (hunk 7 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/feeding_data.rst]: 我想说的本意是 Python中用来做类别标签的数据类型是 :code:`int64` 我改下。

File: source/user_guides/howto/prepare_data/feeding_data.rst (hunk 11 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/feeding_data.rst]: TODO吧，这个写起来有点麻烦，用的人也不多。

Comment (reyoung) [reply source/user_guides/howto/prepare_data/feeding_data.rst]: related #17

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst (hunk 15 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/use_recordio_reader.rst]: 这里没有真的`重排buffer size 大小的数据，进行训练`。是

1. 读入 buffer size大小的数据
2. 随机不重复的返回这个buffer中的每一条数据。

因为 `重排buffer size 大小的数据` 有很多内存操作，很慢。而随机不重复的返回数据会快很多。

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst (hunk 16 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/use_recordio_reader.rst:94]: https://en.wikipedia.org/wiki/Multiple_buffering

我贴个连接吧。。

File: source/user_guides/howto/prepare_data/index.rst (hunk 1 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/index.rst]: Done

File: source/user_guides/howto/prepare_data/feeding_data.rst (hunk 5 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/feeding_data.rst]: Done

File: source/user_guides/howto/prepare_data/feeding_data.rst (hunk 6 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/feeding_data.rst]: Done

File: source/user_guides/howto/prepare_data/feeding_data.rst (hunk 8 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/feeding_data.rst]: Done

File: source/user_guides/howto/prepare_data/feeding_data.rst (hunk 9 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/feeding_data.rst]: Done

Comment (reyoung) [reply source/user_guides/howto/prepare_data/feeding_data.rst]: Done

File: source/user_guides/howto/prepare_data/feeding_data.rst (hunk 10 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/feeding_data.rst]: Done.

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst (hunk 12 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/use_recordio_reader.rst]: Done

**Comment by reyoung** [comment]
@shanyi15  关于PythonReader的问题，我觉得他应该是和『准备数据』相关的topic。不过这篇文章不在我们一开始的规划里。。可以先放到『准备数据』里。

现在侧边栏是这样的

default

剩下的按照意见修改好了。

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst (hunk 16 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/use_recordio_reader.rst:94]: Done.

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst (hunk 15 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/use_recordio_reader.rst]: Done.

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst (hunk 13 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/use_recordio_reader.rst]: Done.

File: source/user_guides/howto/prepare_data/use_recordio_reader.rst (hunk 14 above)
Comment (reyoung) [reply source/user_guides/howto/prepare_data/use_recordio_reader.rst]: Done

Comment (reyoung) [reply source/user_guides/howto/prepare_data/use_recordio_reader.rst]: Done

**Comment by reyoung** [comment]
@guochaorong  Done.

**Review by shanyi15 — APPROVED** [review]
LGTM.
COOL!

**Comment by guochaorong** [comment]
sorry，刚看到， http://ali.reyoung.me:41080/user_guides/howto/prepare_data/index.html 这个地址不能看啦？~
